### ISSUE TITLE
Loss normalization in GRPOTrainer

### ISSUE THREAD

**Issue by YBenPan** [huggingface/trl#2995]
When beta = 0 and num_iterations = 1 in GRPO, the loss should be exactly 0 by doing the math. However, I'm getting non-zero loss, and I think it's related to line 950 in grpo_trainer.py:

`        loss = (per_token_loss * completion_mask).sum() / completion_mask.sum()`

In this case, we are normalizing the advantages across the entire batch of sequences (with potentially different lengths). This is a different formula from the original GRPO algorithm, where we only normalize across each individual sequence. 

I changed the line to: 
`        loss = ((per_token_loss * completion_mask).sum(dim=1) / completion_mask.sum(dim=1)).mean()`
and I got a constant loss of zero, which is correct.

**Comment by qgallouedec** [comment]
Indeed. This comes from https://github.com/huggingface/trl/pull/2881.
Our idea on this is, unless we find that it gives worst results, we should align with classsical loss normalization (global).

Have you compared the two options? If so, the results would be very useful.

**Comment by tchang1997** [comment]
On a related note, is there a reason why the per token loss is globally normalized (L950 of [`grpo_trainer.py`](https://github.com/huggingface/trl/blob/7442d42c21697fd6c0998a75e7478ed4b40490be/trl/trainer/grpo_trainer.py)), but the KL term continues to use per-sequence normalization (L956 of [`grpo_trainer.py`](https://github.com/huggingface/trl/blob/7442d42c21697fd6c0998a75e7478ed4b40490be/trl/trainer/grpo_trainer.py))? 

Looks like the [GRPO paper (Eq. 3)](https://arxiv.org/pdf/2402.03300) sequence-normalizes both (expanding the KL divergence term), so I wonder these should be consistent (i.e., both global normalization, or both sequence-norm, not a mix).

**Comment by qgallouedec** [comment]
Actually L956 is just logging. But you're right that it should use a consistent normalization. Would you like to open a PR to fix this line?

**Comment by tchang1997** [comment]
Ah, I see that now. Anyway — opened a [PR as discussed](https://github.com/huggingface/trl/pull/3004)!
